src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedState.cs

```diff
@@ -122,7 +122,7 @@ public static bool TryGetStateMachineType (MethodDefinition method, [NotNullWhen
 				return null;
 
 			// Avoid repeat scans of the same type
-			if (_cachedTypeToCompilerGeneratedMembers.ContainsKey (type))
+			if (!_cachedTypeToCompilerGeneratedMembers.TryAdd (type, null))
 				return type;
```

Comment:
Just to clarify (might be worth a comment): If we ever get into the recursive call situation, this effectively ignores anything in the compiler generate type (as if it's empty) -> for example suppressions would not work.
I think it's perfectly OK as a behavior, but might be worth a comment to make it clear it was an intentional behavioral choice.